network/f5/bigip_gtm_virtual_server.py

```diff
@@ -0,0 +1,215 @@
+    api = bigsuds.BIGIP(hostname=bigip, username=user, password=password)
+    return api
+
+def virtual_server_exists(api, name, server):
+    # hack to determine if virtual server exists
+    result = False
+    try:
+        virtual_server_id = {'name': name, 'server': server}
+        api.GlobalLB.VirtualServerV2.get_object_status([virtual_server_id])
+        result = True
+    except bigsuds.OperationFailed, e:
+        if "was not found" in str(e):
+            result = False
+        else:
+            # genuine exception
+            raise
+    return result
+
+def add_virtual_server(api, virtual_server_name, virtual_server_server, address, port):
+    addresses = {'address': address, 'port': port}
+    virtual_server_id = {'name': virtual_server_name, 'server': virtual_server_server}
+    api.GlobalLB.VirtualServerV2.create([virtual_server_id], [addresses])
+
+def remove_virtual_server(api, virtual_server_name, virtual_server_server):
+    virtual_server_id = {'name': virtual_server_name, 'server': virtual_server_server}
+    api.GlobalLB.VirtualServerV2.delete_virtual_server([virtual_server_id])
+
+def get_virtual_server_state(api, name, server):
+    virtual_server_id = {'name': name, 'server': server}
+    state = api.GlobalLB.VirtualServerV2.get_enabled_state([virtual_server_id])
+    state = state[0].split('STATE_')[1].lower()
+    return state
+
+def set_virtual_server_state(api, name, server, state):
+    virtual_server_id = {'name': name, 'server': server}
+    state = "STATE_%s" % state.strip().upper()
+    api.GlobalLB.VirtualServerV2.set_enabled_state([virtual_server_id], [state])
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            server = dict(type='str', required=True),
+            user = dict(type='str', required=True),
+            password = dict(type='str', required=True),
+            state = dict(type='str', required=True, choices=['present', 'absent', 'enabled', 'disabled']),
+            host =  dict(type='str', aliases=['address']),
+            port = dict(type='int'),
+            virtual_server_name = dict(type='str', required=True),
+            virtual_server_server = dict(type='str', required=True)
+        ),
+        supports_check_mode=True
+    )
+
+    if not bigsuds_found:
+        module.fail_json(msg="the python bigsuds module is required")
+
+    server = module.params['server']
+    user = module.params['user']
+    password = module.params['password']
+    virtual_server_name = module.params['virtual_server_name']
+    virtual_server_server = module.params['virtual_server_server']
+    state = module.params['state']
+    address = module.params['host']
+    port = module.params['port']
+
+    result = {'changed': False}  # default
+
+    try:
+        api = bigip_api(server, user, password)
+
+        if not virtual_server_exists(api, virtual_server_name, virtual_server_server):
+            module.fail_json(msg="virtual server does not exist")
```

Comment:
Why is this here? Module fails if the virtual server does not exist. But later on you do similar checks depending on when it should be created etc.